### DISCUSSION TITLE
How to get values of dual variables / costates/ lagrange multipliers

### DISCUSSION THREAD

**Discussion by mcgarnett** [coin-or/Ipopt#796] [Q&A]
Hi all, 

These values are important when doing trajectory optimization for scaling and balancing the problem. Poorly scaled problems don't converge. I was wondering if it is possible to read out the values of dual variables with IPOPT, and what function I should look at.  I am using IPOPT through OpenMDAO via PyOptSparse, so I am pretty far removed from the actual IPOPT C++ API. I will have to add this ability to PyOptSparse because I don't see any indication of it there. 

Any pointers or clues on doing this, where should I start looking?

**Comment by svigerske** [comment]
Ipopt returns the values of the dual values to the user in the [finalize_solution callback](https://coin-or.github.io/Ipopt/classIpopt_1_1TNLP.html#a3debb40a1dc203d3d53a93ccc7ea928d). lambda are the duals for the constraints, z_L and z_U the duals for the variable bounds.

**Reply by mcgarnett** [reply]
thank you Stefan
